### ISSUE TITLE
circular serf deadlock

### ISSUE THREAD

**Issue by sternenzaehler** [freeserf/freeserf#462]
Sometimes the serfs would enter a deadlock when for example a stream of outgoing serfs from the castle at some point runs into another stream of serfs who run to the castle by a different route. In this case, if there are enough serfs, the head serf in the "instream" will wait at the castle gate for the outgoing serf to clear the path. But the outgoing serf will wait for his stream to move, that in turn is waiting for the instream to move. 

this case cannot be solved by the game itself, you need to delete one of the roads and rebuild it later.

maybe it would be a smart to detect such cycles and move all the serfs in the cycle by one cell at a time? 

I am not sure but I think this issue exists in the original game as well and is not solved there.

**Comment by tlongstretch** [comment]
Hmm, I ran into this often when testing my AI, I assumed it was a problem with my code.   You are seeing it playing solo?

My work-around is for every AI loop check for serfs with "waiting" attribute, set a timer for it, and once the timer expires (after some few thousand ticks) they are booted off the road by setting their state to "lost" so they wander around off road to the nearest flag.  This fixes it in almost every case without having to destroy roads.  Booting the single serf causing the jam usually frees everybody queued up behind them.

**Comment by tlongstretch** [comment]
I never saw this in the original game, as far as I can remember.  I believe it is a Freeserf bug.

**Comment by tlongstretch** [comment]
Unfortunately, I believe booting stuck serfs leads to another problem where buildings are never staffed, I assume because the booted serf was a professional en route to their work building, and the game never notices they never arrived.  This is pretty rare though because the vast majority of serfs on roads are transporters, and I have yet another check that summons transporters where they are never arrive.  I think I could check to see what job booted serf has and log it so I can see if this is in fact the cause of that condition.

**Comment by sternenzaehler** [comment]
Wow, I never knew there was this code there! I will inspect it to compare it with my findings.

Yes, I a ran into this issue while playing 4 players at a time - not really solo)) I was testing mass operations with many serfs at once.  I ran into the issue while arranging battles - when knights outgoing from the castle ran into the numerous knights streaming into it along all possible paths returning from battles. 

The problem with your fix - to boot a random serf from a road - the fix itself is technically perfect. But the problem is it is not fair (in my view, that is) :). You see, the game may and should run into very long locks due to heavy congestions. In this case booting serfs automatically is not fair, it is cheating. you have to wait until the path clears. Your fix does not distinguish between a deadlock and a heavy congestion lock which may last many minutes at normal game speeds (when a path of serfs intersects with some heavy transport ops especially near mountainous areas). In this latter case it is fair to wait as long as it takes - game rules dictate that. 
But this former case = circular dependence - the game cannot resolve it ever (as I saw). 
I will try to reproduce the issue, save the game and debug this circular resolution code you showed.

**Comment by tlongstretch** [comment]
My "fix" I believe only detects deadlocks that will never resolve themselves, caused by what looks like a bug.  This first stuck serf causing the jam doesn't appear to be waiting on anything legitimate, they are simply stuck on their own for no good reason.  Serfs that are stuck waiting on actual game events (as determined by the result of the serf->is_waiting bool function)  do not have this same WAIT_IDLE_ON_ROAD_ state despite the fact that they may indeed be waiting on a road.  I set my debugging code up to color code serfs affected by each state, and only the bugged serf causing the jam has the bad state.   

code that identifies serfs waiting on game events, these should not be booted as they will resolve themselves if conditions met

serf.cc
`/* Return true if serf is waiting for a position to be available.
   In this case, dir will be set to the desired direction of the serf,
   or DirectionNone if the desired direction cannot be determined. */
`

**Comment by sternenzaehler** [comment]
Yes, this is a different bug. I have seen it, too. a serf is waiting on the road for no legitimate visible reason. I suspected that this is due to the fact that map still has a stale index at that pos. It thinks there is a serf ahead but it is not there. I have seen map become desynchronized with actual serf->pos in large battles so maybe this happens also in the game.

We will find this, pinpoint the exact piece of code and submit a fix:) I am sure.

**Comment by tlongstretch** [comment]
sternenzaehler, is this the issue you are describing?  I just had this happen in a game.  I set up auto-saving every so many ticks so I have a before and during save.   However, if I load the "before" save the deadlock doesn't appear on its own.  I believe it is a timing issue triggered by player action (sending serfs out) and not something easily reproducible this way.

![freeserf_circular_serf_deadlock_issue](https://user-images.githubusercontent.com/45954803/71545653-d2aa8080-2985-11ea-9823-80c6c6b8a7e7.PNG)



[savegame_deadlock_Example.zip](https://github.com/freeserf/freeserf/files/4007146/savegame_deadlock_Example.zip)

**Comment by sternenzaehler** [comment]
Hi tlongstretch, 
Interesting, I am able to reproduce the bug just fine. It requires either sending out an attack party or just trigger a knight exchange for training. The crucial thing is to have a stream of people running to the castle and from the castle - then a deadlock is almost guaranteed on a proper road network ( it wont happen on a simple degenerated star shaped network, but it will on a real complex one). 

I never thought that if you reloaded the deadlock would not happen.
I will try to reproduce it so that I a have a save just before the deadlock that actually works. 
Thanks for your idea!

**Comment by tlongstretch** [comment]
I am seeing what you mean now, I have a savegame where this almost always deadlocks shortly after load.  Sometimes it doesn't, though, I think it has to do with the AI randomness and timing made shortly after loading.
![circular serf deadlock image](https://i.imgur.com/KYbKyDh.png)

**Comment by nicymike** [comment]
Hey @tlongstretch. To fix this you need to change two parts of Serf::handle_serf_walking_state_waiting():

Change the bottom to (wait_counter=0 at wrong position):
```
      dir = (Direction)(other_serf->s.walking.dir + 6);
    }
 /* Wait counter should only be reset inside the if. If not it will never be increased above one. */
  s.walking.wait_counter = 0;
  }

  /* Stick to the same direction */
  change_direction((Direction)(s.walking.dir + 6), 0);
}
```

And change the top if to (the if is messed up):
```
/* Only check for loops once in a while. */
  s.walking.wait_counter += 1;

/* The original code does a revert check:
  if ((map->has_flag(pos) && s.walking.wait_counter < 10) ||
      (!map->has_flag(pos) && s.walking.wait_counter < 50)) {
 ... do nothing here
} else{
 ... check for loop here
}

Basically if a serf is at a flag just count to 10 to prioritize him because the chance is higher that this will solve the deadlock.
If a serf is not at at flag count to 50.
*/
  if ((!map->has_flag(pos) || s.walking.wait_counter >= 10) &&
      (map->has_flag(pos) || s.walking.wait_counter >= 50)) {
    MapPos pos_ = pos;

    /* Follow the chain of serfs waiting for each other and
       see if there is a loop. */

```

I hope this helps.

**Comment by tlongstretch** [comment]
Thank you @nicymike!  I will add this to my fork

**Comment by tlongstretch** [comment]
I think there was another issue causing the circular gridlock detection to never work:

      /* Get next serf and follow the chain */
      //Serf *other_serf = game->get_serf_at_pos(pos);  shouldn't this be pos_ not pos???
      Serf *other_serf = game->get_serf_at_pos(pos_);

Changing pos to pos_ (the next pos) causes it to resolve gridlocks.  Otherwise it never detects them because it breaks as soon as the direction changes (becuase it is looking at the original serf's direction and not the next serf in line)
